Consistent `BedEntry.unpack`

dievsky opened the pull request:
See #44 .

dievsky committed:
BedEntry.unpack now throws on non-conforming receivers

As required by #44. Also expanded kdoc for various classes and methods.

dievsky committed:
Edited BedEntryTest to test for the new behaviour

Also added new tests

iromeo commented on src/main/kotlin/org/jetbrains/bio/big/Bed.kt:
```diff
@@ -66,67 +78,109 @@ data class BedEntry(
+            rest.isEmpty() -> emptyList<String>()
+            omitEmptyStrings -> Splitter.on(delimiter).trimResults().omitEmptyStrings().limit(limit).split(rest)
+            else -> rest.split(delimiter, limit = limit)
+        }.toList()
```
Do we really need to create an intermediate array list for each bed entry? The previous impl just used iterator API.

Also 1st and 3rd `when` branches already return a list, no need to convert to list again.

dievsky replied:
OK, then how about we:
1. replace the third branch with another Splitter chain call (otherwise there's not much sense in demanding an iterator, since a list will be created in most cases anyway);
2. switch to an iterator to prevent the unnecessary list generation, and
3. check `hasNext` for each separate field.

This will require nine more `throw` points, but we'll probably get improved performance over my list-based version.

iromeo replied:
We could at least convert 2nd to list, and leave 1st and 3rd as is (already lists). I think if we use list-based version there is no much sense to convert it to iterator and use `iter.next()` instead of `list[idx]`

P.S: Not sure about any real profit of using iterators here.

dievsky replied:
Done.

iromeo commented on src/main/kotlin/org/jetbrains/bio/big/Bed.kt:
```diff
@@ -66,67 +78,109 @@ data class BedEntry(
             else -> 0
         }
-        val strand = if (fieldsNumber >= 6 && it.hasNext()) it.next().first() else '.'
+        val strand = if (fieldsNumber >= 6) it.next().firstOrNull() ?: '.' else '.'
```
` '.' else '.'` could we avoid this code duplication?

dievsky replied:
Bothers me too, but it's tricky. These dots are somewhat different. One `'.'` is a result of a missing field (`fieldsNumber < 6`), the other requires that `fieldsNumber >= 6`, advances the iterator and receives an empty string from it. I can't think of a nice way to do it.

iromeo replied:
ok

iromeo commented on src/test/kotlin/org/jetbrains/bio/big/BedEntryTest.kt:
```diff
@@ -132,8 +132,35 @@ class BedEntryTest {
+            assertEquals(3, e.fieldIdx, "Missing field index is wrong")
+            return
+        }
+        assertTrue(false, "bed3 parsed as bed12 without throwing")
```
Please use `@Test(expected = BedEntryUnpackException::class)` or 
```
class FooTest {
@Rule @JvmField var thrown = ExpectedException.none()

 @Test fun testSmthFails() {
        smth.init()
        thrown.expect(IllegalArgumentException::class.java)
        thrown.expectMessage("Smth went wrong at 3-rd arg")
        smth.doDangerThing()
    }
}
```

dievsky replied:
Thanks for the suggestion! Done.

iromeo commented on src/main/kotlin/org/jetbrains/bio/big/Bed.kt:
```diff
@@ -53,8 +57,16 @@ data class BedEntry(
     /**
-     * Parses minimal representation
-     * @param fieldsNumber Expected BED format fields number to parse (3..12)
+     * Unpacks a basic bed3 [BedEntry] into a bedN+ [ExtendedBedEntry].
```
Please update https://github.com/JetBrains-Research/big/blob/master/CHANGES. This fix breaks backward compatibility, so we need to inc major version.

dievsky replied:
I'd argue that it doesn't really break compatibility. Earlier we threw an unspecified exception in an unspecified set of cases. Now we throw a specific exception in specific cases.

iromeo replied:
Before the fix, if a user has fewer fields in the file we will get default values for missing fields so we could open any file as `bed12+`. Now he gets an exception in this case.

dievsky replied:
Not quite true. We couldn't open `chr1 1 100 a b` as `bed12+` because `b` would cause a `NumberFormatException` (which by the way wasn't documented anywhere, if I recall correctly). And this forgiving behaviour itself (filling the default values) wasn't documented either.
Either way, I'm totally fine with a major release, I just want to make a (formal) point that we haven't really broken anything that was previously _promised_.

iromeo replied:
ok

iromeo commented on src/main/kotlin/org/jetbrains/bio/big/Bed.kt:
```diff
@@ -66,67 +78,109 @@ data class BedEntry(
+                    0
+                } else {
+                    chunk.toIntOrNull()
+                            ?: throw BedEntryUnpackException(this, 4, "score value $chunk is not an integer")
```
Please add a test for this exception

iromeo commented on src/main/kotlin/org/jetbrains/bio/big/Bed.kt:
```diff
@@ -66,67 +78,109 @@ data class BedEntry(
+                    0
+                } else {
+                    chunk.toIntOrNull()
+                            ?: throw BedEntryUnpackException(this, 6, "thickStart value $chunk is not an integer")
```
Please add a test for this exception

iromeo commented on src/main/kotlin/org/jetbrains/bio/big/Bed.kt:
```diff
@@ -66,67 +78,109 @@ data class BedEntry(
+                    0
+                } else {
+                    chunk.toIntOrNull()
+                            ?: throw BedEntryUnpackException(this, 7, "thickEnd value $chunk is not an integer")
```
Please add a test for this exception

iromeo commented on src/main/kotlin/org/jetbrains/bio/big/Bed.kt:
```diff
@@ -66,67 +78,109 @@ data class BedEntry(
+                    val chunks = value.split(',', limit = 3)
+                    Color(chunks[0].toInt(), chunks[1].toInt(), chunks[2].toInt()).rgb
+                } catch (e: Exception) {
+                    throw BedEntryUnpackException(this, 8, "color value $value is not a comma-separated RGB", e)
```
Please add a test for this exception

iromeo commented on src/main/kotlin/org/jetbrains/bio/big/Bed.kt:
```diff
@@ -66,67 +78,109 @@ data class BedEntry(
+                    0
+                } else {
+                    chunk.toIntOrNull()
+                            ?: throw BedEntryUnpackException(this, 9, "blockCount value $chunk is not an integer")
```
Please add a test for this exception

iromeo commented on src/main/kotlin/org/jetbrains/bio/big/Bed.kt:
```diff
@@ -66,67 +78,109 @@ data class BedEntry(
+                try {
+                    value.splitToInts(blockCount)
+                } catch (e: Exception) {
+                    throw BedEntryUnpackException(
```
Please add a test for this exception

iromeo commented on src/main/kotlin/org/jetbrains/bio/big/Bed.kt:
```diff
@@ -66,67 +78,109 @@ data class BedEntry(
+                try {
+                    value.splitToInts(blockCount)
+                } catch (e: Exception) {
+                    throw BedEntryUnpackException(
```
Please add a test for this exception

iromeo commented:
Please ensure that all tests are green, according to Travis CI, `BigBedFileTest` test fails.

dievsky committed:
Edited BigBedFileTest to conform to the new behaviour

dievsky committed:
Added ExpectedException approach to BedEntryTest

dievsky committed:
Replaced iteration with get in BedEntry.unpack

dievsky committed:
Better extra fields splitting in BedEntry.unpack

We split everything just once

dievsky committed:
Replaced extraFieldsNumber with parseExtraFields flag in BedEntry.unpack

We actually only used 0 or null as extraFieldsNumber, which correspond
to the two values of the newly introduced flag.
Also updated BedEntryTest accordingly.

dievsky committed:
More strict splitToInts

Will throw if there are too few elements in the string

dievsky committed:
Added eight new tests to BedEntryTest

They test for eight new "throw"s in BedEntry.unpack

dievsky committed:
Added a deprecated unpack with old signature for compatibility

dievsky committed:
Added CHANGES description

iromeo commented on src/main/kotlin/org/jetbrains/bio/big/Bed.kt:
```diff
@@ -59,35 +59,35 @@ data class BedEntry(
     fun unpack(
             fieldsNumber: Byte = 12,
-            extraFieldsNumber: Int? = null,
+            parseExtraFields: Boolean = true,
```
Seems it is inconsistent with `filedsNumber`. We validate bed fields number and throw exceptions if number differs from expected. On the other hand now we are "relaxed" in terms of extra fields number, any number allowed. Is it intended behavior?

dievsky replied:
It is.
I posit that, since extra fields are extra anyway, and we give them to the user as-is, there's no point in imposing strict limitations on them. Any semantics that they might have is the user's responsibility. The regular fields are another story -- we parse them. If the user specifies `bed9`, they must really want that color field, and it would be misleading to just fill it with default value.
Another point is that within my implementation, once you know the delimiter, the set of formats becomes strictly ordered: the sets of `BedEntry`s recognizable by any two formats are nested. Any `BedEntry` can be unpacked with `bed3+`, any `bed7+` entry is simultaneously a `bed5+` entry etc. This is important because under this condition you can determine a common format for any file.

iromeo replied:
> , since extra fields are extra anyway, and we give them to the user as-is

I don't agree.

Pure BED format is not widely used by tools, popular peak callers used own fixed bedN+K format (macs, sicer, span) with fixed `k`. 

If we expected that our format is bed3+5, than we assume that we could use extra fields 0..4 and won't get ArrayOutOfBounds exception, or etc. In case of bed3+ we indeed could don't care of extra fields number

> If the user specifies bed9, they must really want that color field, and it would be misleading to just fill it with default value.

The same idea here, if the user expects 'K' extra fileds it would misleading to get AIOBE when accessing to k-th field. Why he is required to check each time that `K` fileds were parsed indeed(we parse them as string array) if he already asked to parse `k` fields?

dievsky replied:
See, in my current version, the user simply can't specify `bed3+5`, because there's no such thing. So it's consistent and sufficiently powerful (`bedN+K` format is a subset of `bedN+` format).
But we can revert this decision and return to the support of all three versions (`bedN+`, `bedN+K` and `bedN` as a special case of `bedN+K` with `K == 0`) if you believe that this is a viable use case. Should we throw a separate exception on missing extra fields or just let the user figure where the error lies?

iromeo replied:
> if you believe that this is a viable use case. 

@dievsky So let's leave it as you like. My point was that there are some new inconsistency of `bedN+K` vs `bedN` parsing (relaxed for K and strict for N), I'm not sure that it is a critical or viable use case, we could return to this in future if smth goes wrong.

iromeo commented on src/main/kotlin/org/jetbrains/bio/big/Bed.kt:
```diff
@@ -211,12 +215,9 @@ data class BedEntry(
-        check(s.hasNext() == (size > 0))
-        while (s.hasNext() && ptr < size) {
+        while (ptr < size) {
             chunks[ptr++] = s.next().toInt()
```
Not clear what for is `More strict splitToInts`. Also if `s` has fewer chunks then `size` this code will throw an AOBE?

I don't remember the exact use case, probably is when a user needs rgb-color and only one array based column, e.g. (block sizes or block starts), so uses '0' or '.' for the unused column.

dievsky replied:
Standard conformance, mostly:
> blockSizes - A comma-separated list of the block sizes. The number of items in this list should correspond to blockCount.

We can revert this change, I don't have any particularly strong feelings for it. Just say the word.
Yes, `IOOBE` will be thrown. It will also be immediately caught at both call sites of this private function.
`color` is column 9, while `blockStarts` and `blockSizes` are 12 and 11 respectively, so the use case must be different.

iromeo replied:
Ok, let's leave as is until we've faced some real issue

dievsky committed:
Edited CHANGES description

iromeo commented on CHANGES:
```diff
@@ -3,13 +3,18 @@ big Changelog
 Here you can see the full list of changes between each big release.
 
-Version 0.8.5
+Version 0.9.0
```
Don't forget to update version in `gradle.properties`

dievsky replied:
Done.

dievsky committed:
Edited CHANGES description

Ready for 0.9.0 release.

dievsky committed:
Bumped version in gradle.properties
